BWS-78: + backend and frontend

ania-leichenko opened the pull request:
![Screenshot_6](https://user-images.githubusercontent.com/65825815/156141239-c6b609f6-6a81-475d-a2ef-fdd345055784.png)

what1s1ove commented on frontend/src/components/eam/components/tenant/tenant.tsx:
```diff
@@ -0,0 +1,84 @@
+  EAMTenantFormDto,
+} from 'common/types/types';
+import { eamTenantUpdate as EAMTenantValidationSchema } from 'validation-schemas/validation-schemas';
+import { app as EAMTenantConfigurateActions } from 'store/actions';
```
this is part of the `eam` page. so move all actions there

what1s1ove commented on frontend/src/components/eam/components/tenant/tenant.tsx:
```diff
@@ -0,0 +1,84 @@
+    tenant: app.tenant,
+  }));
+
+  const defaultValues = useMemo(() => {
+    const values = {
+      ...DEFAULT_PAYLOAD,
+    };
+    if (tenant) {
+      values.name = tenant.name;
+    }
+    return values;
+  }, [tenant]);
```
```suggestion
  const defaultValues = useMemo(() => mapTenantDefaultValues(tenant), [tenant]);
```
create this helper `mapTenantDefaultValues`

what1s1ove commented on frontend/src/components/eam/components/tenant/tenant.tsx:
```diff
@@ -0,0 +1,84 @@
+  }, [tenant]);
+
+  const handleFormSubmit = (payload: EAMTenantFormDto): void => {
+    if (tenant) {
+      dispatch(
+        EAMTenantConfigurateActions.updateTenant({
+          ...payload,
+          id: tenant.id,
+        }),
+      );
+    }
```
```suggestion
      dispatch(
        EAMTenantConfigurateActions.updateTenant({
          ...payload,
          id: (tenant as Tenant).id,
        }),
      );
```

ania-leichenko replied:
where do we get this Tenant?

what1s1ove replied:
I mean a type that we store in the redux store

what1s1ove commented on frontend/src/hooks/use-app-form/use-app-form.hook.ts:
```diff
@@ -1,4 +1,9 @@
+  UseFormHandleSubmit,
+  UseFormReset,
+  FieldValues,
+} from 'react-hook-form';
```
use type from our types folder

ania-leichenko replied:
Do you want me to transfer types?

what1s1ove replied:
Only that we don't have
most of them we already have

what1s1ove commented on frontend/src/hooks/use-app-form/use-app-form.hook.ts:
```diff
@@ -16,6 +21,7 @@ type UseAppFormResult<T extends FormControlValues = FormControlValues> = {
   control: FormControl;
   errors: FormControlErrors;
   handleSubmit: UseFormHandleSubmit<T>;
+  reset: UseFormReset<FieldValues>;
```
```suggestion
  handleReset: UseFormReset<FieldValues>;
```

what1s1ove commented on frontend/src/hooks/use-app-form/use-app-form.hook.ts:
```diff
@@ -37,6 +44,7 @@ const useAppForm = <T extends FormControlValues = FormControlValues>({
     control,
     errors,
     handleSubmit: handleSubmit as UseFormHandleSubmit<T>,
+    reset,
```
```suggestion
    handleReset: reset,
```

what1s1ove commented on frontend/src/services/tenant-api/tenant-api.service.ts:
```diff
@@ -31,6 +32,19 @@ class TenantApi {
+    params: EAMTenantUpdateRequestDto,
+  ): Promise<EAMTenantByIdResponseDto> {
+    return this.#http.load(
+      joinItems(this.#apiPrefix, ApiPath.TENANTS, '/', params.id),
```
```suggestion
      joinItems(this.#apiPrefix, ApiPath.TENANTS, TenantsApiPath.ROOT, params.id),
```

what1s1ove commented on frontend/src/store/app/actions.ts:
```diff
@@ -18,4 +19,15 @@ const getTenant = createAsyncThunk<
+>(ActionType.UPDATE_TENANT_NAME, async (payload, { extra }) => {
+  const { tenantApi, notification } = extra;
+  const tenant = await tenantApi.updateTenant(payload);
+  notification.success('Success!', 'Tenant has been successfully created');
```
we have enums for it. use them

ania-leichenko replied:
What do you mean?, I don't quite understand, tell me please

what1s1ove replied:
Check an example - https://github.com/BinaryStudioAcademy/bsa-winter-2021-2022-bws/blob/development/frontend/src/store/eam-group-configurate/actions.ts#L35

ania-leichenko replied:
thanks)

what1s1ove commented on frontend/src/store/app/actions.ts:
```diff
@@ -18,4 +19,15 @@ const getTenant = createAsyncThunk<
 });
 
-export { getTenant };
+const updateTenant = createAsyncThunk<
```
action should be in the `eam` page scope

what1s1ove commented on frontend/src/store/app/common.ts:
```diff
@@ -1,5 +1,6 @@
 enum ActionType {
   GET_TENANT = 'app/get-tenant',
+  UPDATE_TENANT_NAME = 'eam-configurate-tenant/update-tenant-name',
```
```suggestion
```

what1s1ove commented on shared/src/validation/eam-tenant-update/eam-tenant-update.validation-schema.ts:
```diff
@@ -0,0 +1,23 @@
+    .trim()
+    .min(EAMTenantValidationRule.NAME_MIN_LENGTH)
+    .max(EAMTenantValidationRule.NAME_MAX_LENGTH)
+    // .regex(EAMTenantValidationRule.NAME_REGEX)
```
remove comments

what1s1ove reviewed (changes requested):
<img width="1375" alt="image" src="https://user-images.githubusercontent.com/42296182/155959720-6c480e70-497f-4579-8fda-bd3ac794e5ad.png">
lets add title. something like "Change Tenant Name" as title and "Tenant Name" as label for input

Yaroslaww-1 commented on backend/src/data/repositories/tenant/tenant.repository.ts:
```diff
@@ -26,6 +26,17 @@ class Tenant {
     return Tenant.modelToEntity(tenant);
   }
 
+  async updateById(id: string, name: string): Promise<TenantEntity | null> {
```
Repositories should not contain `updateById`-like methods. Here is a quick explanation:
Often when we need to update something we also decide to have some validation before updating. After some time these validations overwhelm repository methods and make them unsopportable.
If you want update data you need to create update methods on Entities and add `save(tenant: TenantEntity)` method to repository

Yaroslaww-1 commented on backend/src/services/tenant/tenant.service.ts:
```diff
@@ -32,6 +32,25 @@ class Tenant {
+    id: string;
+    name: string;
+  }): Promise<EAMTenantByIdResponseDto | null> {
+    const tenant = await this.#tenantRepository.updateById(id, name);
+
+    if (!tenant) {
+      return null;
+    }
```
```suggestion
   const tenant = await repo.getById(id);
   if (!tenant) ...
   tenant.setName(name)
   await repo.save(tenant)
```

ania-leichenko committed:
BWS-78: + backend and frontend

BWS-78: * review fixes

BWS-78: * fix validation

BWS-78: * fix problems

BWS-78: * backend

BWS-78: + error message

what1s1ove commented on frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts:
```diff
@@ -0,0 +1,16 @@
+import { EAMTenantFormDto, EAMTenantByIdResponseDto } from 'common/types/types';
+
+const mapTenantDefaultValues = (
+  tenant: EAMTenantByIdResponseDto | null,
```
```suggestion
  tenant: EAMTenantByIdResponseDto,
```

what1s1ove commented on frontend/src/components/eam/components/tenant/tenant.tsx:
```diff
@@ -0,0 +1,76 @@
+    tenant: app.tenant,
+  }));
+
+  const defaultValues = useMemo(() => mapTenantDefaultValues(tenant), [tenant]);
```
```suggestion
  const defaultValues = useMemo(() => mapTenantDefaultValues(tenant as EAMTenantByIdResponseDto), [tenant]);
```

what1s1ove commented on frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts:
```diff
@@ -0,0 +1,16 @@
+    values.name = tenant.name;
+  }
+  return values;
+};
```
```suggestion
): EAMTenantFormDto => ({
  name: tenant.name,
  // other keys 
});
```

ania-leichenko replied:
what if tennant = null?

```jsx
): EAMTenantFormDto => ({
  name: tenant?.name ?? '',
  // other keys 
});
```

what1s1ove commented on frontend/src/components/eam/components/tenant/tenant.tsx:
```diff
@@ -0,0 +1,76 @@
+} from 'common/types/types';
+import { eamTenantUpdate as EAMTenantValidationSchema } from 'validation-schemas/validation-schemas';
+import { eam as EAMTenantConfigurateActions } from 'store/actions';
+import { mapTenantDefaultValues } from './helpers/tenant-default-values.helper';
```
```suggestion
import { mapTenantDefaultValues } from './helpers/helpers';
```

what1s1ove commented on frontend/src/store/eam/common.ts:
```diff
@@ -1,6 +1,7 @@
 enum ActionType {
   GET_WORKERS = 'eam/get-workers',
   LOAD_GROUPS = 'eam/load-groups',
+  UPDATE_TENANT_NAME = 'eam-configurate-tenant/update-tenant-name',
```
```suggestion
  UPDATE_TENANT_NAME = 'eam/update-tenant-name',
```

what1s1ove commented on frontend/src/store/eam/reducer.ts:
```diff
@@ -4,20 +4,23 @@ import { getWorkers, loadGroups } from './actions';
   workers: EAMWorkerGetAllItemResponseDto[];
   groupsDataStatus: DataStatus;
   groups: EAMGroupGetByTenantResponseItemDto[];
+  tenant: EAMTenantByIdResponseDto | null;
```
```suggestion
```
we don't want to keep a tenant here. we should keep a tenant in the `app` reducer, but action should for changing should be in the `eam` scoop

what1s1ove commented on frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts:
```diff
@@ -0,0 +1,9 @@
+const mapTenantDefaultValues = (
+  tenant: EAMTenantByIdResponseDto,
+): EAMTenantFormDto => ({
+  name: tenant ? tenant.name : '',
```
```suggestion
  name: tenant?.name ?? '',
```

what1s1ove commented on frontend/src/components/eam/components/tenant/tenant.tsx:
```diff
@@ -0,0 +1,79 @@
+  }));
+
+  const defaultValues = useMemo(
+    () => mapTenantDefaultValues(tenant as EAMTenantByIdResponseDto),
```
```suggestion
    () => mapTenantDefaultValues(tenant),
```
we discussed that we should cast a type

what1s1ove commented on frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts:
```diff
@@ -0,0 +1,9 @@
+import { EAMTenantFormDto, EAMTenantByIdResponseDto } from 'common/types/types';
+
+const mapTenantDefaultValues = (
+  tenant: EAMTenantByIdResponseDto,
```
```suggestion
  tenant: EAMTenantByIdResponseDto | null,
```

Yaroslaww-1 commented on backend/src/services/tenant/tenant.service.ts:
```diff
@@ -32,6 +34,33 @@ class Tenant {
+    if (!tenant) {
+      return null;
+    }
+    if (tenant.name === name) {
+      throw new TenantError({
+        status: HttpCode.DENIED,
+        message: ExceptionMessage.TENANT_NAME_EXISTS,
+      });
+    }
```
It won't work. Consider an example where we have two tenants:
tenant1 - id1
tenant2 - id2
PUT /id1 with name tenant2
```suggestion
    const tenantWithName = await repo.getByName(name);
    if (tenantWithName !== null) {
      throw new TenantError({
        status: HttpCode.DENIED,
        message: ExceptionMessage.TENANT_NAME_EXISTS,
      });
    }
```

what1s1ove commented on backend/src/services/tenant/tenant.service.ts:
```diff
@@ -32,6 +34,34 @@ class Tenant {
+  }): Promise<EAMTenantByIdResponseDto | null> {
+    const tenant = await this.#tenantRepository.getById(id);
+    if (!tenant) {
+      return null;
```
throw an error here

ania-leichenko committed:
BWS-78: + error message
